crates/bindings-csharp/BSATN.Codegen/Type.cs

```diff
@@ -253,36 +263,47 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                 "SpacetimeDB.TypeAttribute",
-                predicate: (node, ct) =>
-                {
-                    // structs and classes should be always processed
-                    if (node is not EnumDeclarationSyntax enumType)
+                // for enums, we just do diagnostics, nothing else
+                predicate: (node, ct) => node is EnumDeclarationSyntax,
+                transform: (context, ct) =>
+                    context.ParseWithDiags(diag =>
                     {
-                        return true;
-                    }
+                        var enumType = (EnumDeclarationSyntax)context.TargetNode;
+
+                        // Ensure variants are contiguous as SATS enums don't support explicit tags.
+                        foreach (var variant in enumType.Members)
+                        {
+                            if (variant.EqualsValue is { } equalsValue)
+                            {
+                                diag.Report(
+                                    ErrorDescriptor.EnumWithExplicitValues,
+                                    (equalsValue, variant, enumType)
+                                );
+                            }
+                        }
+
+                        // Ensure all enums fit in `byte` as that's what SATS uses for tags.
+                        if (enumType.Members.Count > 256)
+                        {
+                            diag.Report(ErrorDescriptor.EnumTooManyVariants, enumType);
+                        }
+
+                        // Unused empty type.
+                        return default(ValueTuple);
+                    })
+            )
+            .ReportDiagnostics(context)
+            .WithTrackingName("SpacetimeDB.Type.ParseEnum");
 
-                    // Ensure variants are contiguous as SATS enums don't support explicit tags.
-                    if (enumType.Members.Any(m => m.EqualsValue is not null))
-                    {
-                        throw new InvalidOperationException(
-                            "[SpacetimeDB.Type] enums cannot have explicit values: "
-                                + enumType.Identifier
-                        );
-                    }
-
-                    // Ensure all enums fit in `byte` as that's what SATS uses for tags.
-                    if (enumType.Members.Count > 256)
-                    {
-                        throw new InvalidOperationException(
-                            "[SpacetimeDB.Type] enums cannot have more than 256 variants."
-                        );
-                    }
-
-                    // Check that enums are compatible with SATS but otherwise skip from extra processing.
-                    return false;
-                },
-                transform: (context, ct) => new TypeDeclaration(context)
+        context
+            .SyntaxProvider.ForAttributeWithMetadataName(
+                "SpacetimeDB.TypeAttribute",
```

Comment:
Forgot to typeof(TypeAttribute).FullName this one?

Reply:
No, BSATN.Codegen doesn't include BSATN.Runtime/Attrs.cs, at least yet. We don't really have useful attribute info, so I didn't bother sharing it between those packages too, for now that only happens in module's Runtime+Codegen.